### PR TITLE
[None][chore] Remove dead code in model engine

### PR DESCRIPTION
**Pull request by mikeiovine**
## Description

Remove all dead `is_draft_model` branches.

## Test Coverage

Existing tests.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- The current diff modifies six files: `_util.py`, `cuda_graph_runner.py`, `model_engine.py`, `py_executor.py`, `speculative/interface.py`, and one unit-test file.
- The source changes remove draft-model state and simplify related speculative-decoding APIs and execution paths.
- Verify API consistency across KV-cache creation, CUDA graph handling, model-engine forwarding, and speculative decoding.
- No test results or CI failure details were supplied. Review finding counts are unavailable.

## QA Engineer Review

- One unit-test file changed: `tests/unittest/_torch/executor/test_pytorch_model_engine.py`.
- The test changes remove obsolete draft-model setup and update model-engine test configuration.
- No integration test-list files changed, so no test-list update is required.
- Coverage verdict: needs follow-up because no test execution result is supplied.

## Per-File QA Perspective

- `tensorrt_llm/_torch/pyexecutor/_util.py`: Verify KV-cache manager defaults and KV event configuration.
- `tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py`: Verify graph-key generation and draft-length and context validation.
- `tensorrt_llm/_torch/pyexecutor/model_engine.py`: Verify model-engine initialization, input preparation, forwarding, warmup, and speculative execution after draft-path removal.
- `tensorrt_llm/_torch/pyexecutor/py_executor.py`: Verify speculative decoding overlap and acceptance-tensor propagation.
- `tensorrt_llm/_torch/speculative/interface.py`: Verify speculative attention-mode selection without `is_draft_model`.
- `tests/unittest/_torch/executor/test_pytorch_model_engine.py`: Updates model-engine fixtures and draft-model-related cases. This unit-test file is not listed in integration test lists.

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/interface.py [hunk 1]
```diff
@@ -431,22 +431,20 @@ def extend_ctx(self, attention_backend: Type[AttentionBackend]):
...
             attention_backend: the attention backend.
         """
         is_trtllm_attention = issubclass(attention_backend, TrtllmAttention)
 
         # Always use the multi-token query mode for 1-model if the kernels are available.
         use_case_1 = self.use_one_engine()
-        # For 2-model, only the target model (verification) processes multiple tokens at once.
-        use_case_2 = (not self.use_one_engine() and not is_draft_model
-                      and is_trtllm_attention)
+        # For external drafters (NGram, user-provided), the target model verifies
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/speculative/interface.py]: This comment misuses the codebase's own term. `is_external_drafter()` (line 377) means parallel-draft (PARD/DFlash/DSpark) and `DRAFT_TARGET_ONE_MODEL`, and all of those are *inside* `use_one_engine()`, so they hit `use_case_1`, not this branch. The modes that actually reach `use_case_2` are the non-one-engine ones: NGram, user-provided, and the remaining two-model draft/target modes. Suggest:

```python
# For modes that do not run in one engine (NGram, user-provided drafts,
# two-model draft/target), the target model verifies multiple draft
# tokens per step and needs the multi-token query kernel.
```

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 2]
```diff
@@ -5958,8 +5669,7 @@ def forward(self,
                 new_tensors_device: Optional[SampleStateTensors] = None,
                 gather_context_logits: bool = False,
                 cache_indirection_buffer: Optional[torch.Tensor] = None,
-                num_accepted_tokens_device: Optional[torch.Tensor] = None,
-                req_id_to_old_request: Optional[Dict[int, LlmRequest]] = None):
+                num_accepted_tokens_device: Optional[torch.Tensor] = None):
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: `forward` still accepts `num_accepted_tokens_device`, and `PyExecutor._forward_step` still computes it and passes it here, but it is now dropped before `_forward_scheduled`. A parameter the caller feeds and the callee ignores will mislead the next reader into thinking it matters. Since the only consumer was the removed draft-model input rewrite, consider removing it from the `ModelEngine` ABC (line 212), `ad_executor.py:1057`, and the `model_engine.forward(...)` call in `py_executor.py` in this PR. If you'd rather keep the ABC signature stable, at least add a short comment here that it is accepted for interface compatibility and unused. (The drafter's use of the same tensor via `generate_draft_tokens_with_overlap` is still live; only this leg is dead.)

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 3]
```diff
@@ -379,52 +377,37 @@ def get_graph_key(
...
+            if any(
+                    int(request.context_chunk_size) != context_query_len
+                    for request in context_requests[1:]):
+                return None
+        num_encoder_tokens = sum(
+            int(request.encoder_output_len) for request in context_requests
+            if not request.py_skip_cross_kv_projection)
+        return KeyType(batch_size=batch_size,
+                       draft_len=draft_len,
+                       is_first_draft=False,
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:402]: Follow-up, not blocking: with this change `KeyType.is_first_draft` is a constant `False`, `py_is_first_draft` has no remaining setter to `True` (readers at `[cuda_graph_runner.py:321](https://github.com/NVIDIA/TensorRT-LLM/pull/19513/files#diff-04500eda9435bc22d845f9f29171528019dfde8a0f90a53afa51df7c4f21f76dR321)` and `[model_engine.py:175](https://github.com/NVIDIA/TensorRT-LLM/pull/19513/files#diff-203c6491ca0bedf455eeee3ed7939c3209f3f14c383594c26261e1dc48a00395R175)/186/4783` are dead), and `Eagle3ResourceManager.is_first_draft` is now write-only. Worth a second chore PR to remove the field and those branches rather than leaving half of the two-model plumbing behind.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

Premise checks out: `py_executor_creator.py:610` sets `draft_model_engine = None` and never reassigns it, and neither `PyTorchModelEngine(` construction site passes `is_draft_model`, so every removed branch was unreachable. The collateral is that the draft-only overrides (sparse attention off, LoRA off, KV events off, `DRAFT_KV_CACHE_MANAGER` key) collapse to the target-model values, which is a no-op for the same reason. Worth a sentence in the description so a future reader sees why the removal is safe.

Two things inline: the rewritten `attention_need_spec_dec_mode` comment describes the wrong set of modes, and `forward()` still accepts an argument the executor computes but the engine now ignores. Plus one follow-up note on `is_first_draft` remnants that are now write-only.

Not run locally (no torch in my environment); relying on the touched unit tests in CI.

**Review by allisonlim-nv — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 2 above)
Comment (BowenFu) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: +1

**Review by SimengLiu-nv — APPROVED** [review]
LGTM.

**Review by Tabrizian — APPROVED** [review]
No disagg changes, approving.

**Review by lori-ren — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py (hunk 3 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:402]: Yep, will continue to clean this stuff up in follow ups. This one should be removed next.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 2 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Removed

**Review by bmarimuthu-nv — APPROVED** [review]

**Review by govind-ramnarayan — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -2545,7 +2466,7 @@ def _get_graphs_to_capture(
                         f"Capturing {len(graphs)} graphs: {graphs}")
             return graphs
 
-        # Case 3: Target model (two-model) or one-model without dynamic draft
+        # Case 2: One-model without dynamic draft
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:2469]: The new label says "One-model without dynamic draft", but the guard four lines below (`not self.spec_config.spec_dec_mode.use_one_engine()`) only fires for modes that are *not* one-engine, and `spec_config is None` is also handled here. The previous comment ("Target model (two-model) or one-model without dynamic draft") was closer. Suggest: `# Case 2: static draft length (one-model without dynamic draft, or a target model with an NGram/user-provided drafter, or no spec decode)`.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

Both threads from the last round are fixed in the current head, not just resolved:

- `attention_need_spec_dec_mode` comment (`[tensorrt_llm/_torch/speculative/interface.py:450](https://github.com/NVIDIA/TensorRT-LLM/pull/19513/files#diff-d8287cf5ac0bab6a0aec76cc4d149a8f3ecaa3901c90e1157ed826a220a90520R450)-453`) now names NGram and user-provided drafts as the non-one-engine modes, which matches `use_one_engine()` at line 341.
- `num_accepted_tokens_device` is gone from the `ModelEngine` ABC (`[tensorrt_llm/_torch/pyexecutor/model_engine.py:206](https://github.com/NVIDIA/TensorRT-LLM/pull/19513/files#diff-203c6491ca0bedf455eeee3ed7939c3209f3f14c383594c26261e1dc48a00395R206)-211`), from `ADEngine.forward` (`[tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py:1050](https://github.com/NVIDIA/TensorRT-LLM/pull/19513/files#diff-eb8f9dbeda9cd8af984450069e640714f50fb002dc42d1d16d44de34f36beac2R1050)-1057`), and from `PyExecutor._forward_step`. The only remaining use is the local in `_handle_speculative_decoding` that feeds `generate_draft_tokens_with_overlap`, which is a live path.

The third thread (`KeyType.is_first_draft` is now a constant `False`, `py_is_first_draft` has no setter to `True`) was deferred to a follow-up, which is fine. Two more dead leftovers belong on that same follow-up list, non-blocking:

- `get_graph_key` still accepts `spec_resource_manager` (`[tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:350](https://github.com/NVIDIA/TensorRT-LLM/pull/19513/files#diff-04500eda9435bc22d845f9f29171528019dfde8a0f90a53afa51df7c4f21f76dR350)`) and `maybe_get_cuda_graph` still forwards it (line 569), but the removed Eagle3 branch was the only reader inside `get_graph_key`.
- `Eagle3ResourceManager.is_first_draft` (`tensorrt_llm/_torch/speculative/eagle3.py:105,128`) now has no reader in the engine or graph runner.

Description note: besides the `is_draft_model` branches, the diff also removes `num_accepted_tokens_device` / `req_id_to_old_request` from the `forward` chain and the `ModelEngine` ABC. Worth one line in the description so a reader of the merged commit knows the ABC signature changed on purpose.

Verification: no torch in my checkout, so I could not run the unit tests. All 22 changed files byte-compile, and a repo-wide grep finds no remaining engine-level `is_draft_model`, no two-argument `attention_need_spec_dec_mode` caller, and no surviving `Eagle3ResourceManager` reference in the two files that dropped the import. CI is the real check here.

File: tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py
Comment (BowenFu) [inline tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py]: Thanks for putting this together! The 1.3 release branch is about to be cut from main (planned for 9/29), and until then we're asking larger changes and refactors like this one to wait so main stays stable for the branch-out. If this really needs to make 1.3 or a nightly/weekly release, could you explain why in the PR description and in your review request? I'll resolve this thread once that's there, or after branch-out either way. This is only a timing hold, not a concern with the change itself.
